src/ofxstatement/statement.py

```diff
@@ -166,6 +189,50 @@ def generate_transaction_id(stmt_line):
     return h.hexdigest()
 
 
+def generate_unique_transaction_id(stmt_line, unique_id_set):
+    """
+    Generate a unique transaction id.
+
+    A bit of background: the problem with these transaction id's is that
+    they do do not only have to be unique, they also have to stay the same
+    for the same transaction every time you generate the statement.  So
+    generating random ids will not work, even though they will be unique,
+    GnuCash or beancount will recognize these transaction as "new" if you
+    happen to generate and import the same statement twice or import two
+    statements with overlapping periods.
+
+    The function generate_transaction_id() is deterministic, but does not
+    necesserily generate an unique id.
+
+    Therefore this function improves on it since you can create a
+    really unique id by adding an increment to the generated id (a string)
+    and keep on incrementing till it succeeds.
+
+    These are the steps:
+    1) supply a unique id set you want to use for checking uniqueness
+    2) next you generate an initial id by calling
+       generate_transaction_id()
+    3) assign the initial id to the current id (id)
+    4) increment a counter while the current id is a member of the set and
+       add the counter to the initial id and assign that to the current id
+    5) add the current id to the unique id set
+    6) return a list of the current id and the counter
+
+    The counter is returned in order to enable the caller to modify
+    its statement line, for example the memo field.
+    """
+    assert isinstance(unique_id_set, set)
+    # Save the initial id
+    id = initial_id = generate_transaction_id(stmt_line)
+    counter = 0
+    while id in unique_id_set:
+        counter += 1
+        id = initial_id + str(counter)
+
+    unique_id_set.add(id)
+    return [id, counter]
```

Comment:
Okay, this MR looks all good to me except the return value in this function. Right now it doesn't even return the unique id it generates. And the type it returns (a pair) is hard to justify for me - you have to require users to do something with this pair. I'd expect just a unique id returned.

That said I think I understand where you are coming from - you want to detect duplicate statement lines and "adjust" them a bit, if counter is not 0. But that seems very specific to that particular problem with beancount-import don't fully understand. So it is hard to explain why we have chosen this return type instead of more obvious one without explaining the whole story behind this function.

For example, we can generate ids that would look like `{id}-{counter}`, so you can split that generated id in your code to find out the index. Parsing ids is not great, but that would be a compromise.

What do you think?

Reply:
Thanks, I will change accordingly and commit soon. I will also remove the new exception handling code in tool.py but leave the statement.assert_valid().